### PR TITLE
[IMP] stock: scheduler modification

### PR DESCRIPTION
**Pull request by yhu-odoo**
1. When reservation can't be done for some products, we automatically
evaluate their reordering rules and create the corresponding
replenishment document.
2. When receiving products in stock, automatically check if some stock
moves are available and assign them.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_orderpoint.py [hunk 1]
```diff
@@ -240,6 +243,7 @@ def _compute_qty(self):
             for orderpoint in orderpoints_by_context:
                 orderpoint.qty_on_hand = products_qty[orderpoint.product_id.id]['qty_available']
                 orderpoint.qty_forecast = products_qty[orderpoint.product_id.id]['virtual_available'] + products_qty_in_progress[orderpoint.id]
+        self._compute_qty_to_order()
```
Comment (sle-odoo) [inline addons/stock/models/stock_orderpoint.py:246]: not sure about the changes in this file, can you double check @amoyaux ?

Comment (yhu-odoo) [reply addons/stock/models/stock_orderpoint.py:246]: @sle-odoo  @amoyaux . Sorry, it's just for me to check if some issues are caused by the wrong qty_to_order. I'll remove that.

File: addons/stock/models/stock_move.py [hunk 2]
```diff
@@ -1441,6 +1442,13 @@ def _action_done(self, cancel_backorder=False):
         for company_id, move_dests in move_dests_per_company.items():
             move_dests.sudo().with_company(company_id)._action_assign()
 
+        # if incoming moves make other confirmed/partially_available moves available, assign them
+        domains = []
+        for move in moves_todo.filtered(lambda m: m.picking_id and m.picking_id.picking_type_id.code == 'incoming'):
+            domains.append([('product_id', '=', move.product_id.id), ('location_id', '=', move.location_dest_id.id)])
+        moves_to_reserve = self.search(expression.AND([[('state', 'in', ['confirmed', 'partially_available']), ('procure_method', '=', 'make_to_stock')], expression.OR(domains)]))
+        moves_to_reserve._action_assign()
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py:1450]: can you move this block into stock_picking.py action_assign?

File: addons/stock/models/stock_picking.py [hunk 3]
```diff
@@ -680,6 +680,11 @@ def action_assign(self):
         package_level_done.write({'is_done': False})
         moves._action_assign()
         package_level_done.write({'is_done': True})
+
+        # run scheduler for moves are not fully reserved
+        for move in moves.filtered(lambda m: m.state in ['confirmed', 'partially_available']):
+            move.product_id.orderpoint_ids.filtered(lambda op: op.location_id == move.location_id and op.trigger == 'auto').action_replenish(raise_user_error=False)
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py:686]: you can use _procure_orderpoint_confirm instead of action_replenish
you should use op.location_id parent_of move.location_id

File: addons/stock/models/stock_orderpoint.py [hunk 4]
```diff
@@ -209,8 +209,11 @@ def write(self, vals):
     def action_open_orderpoints(self):
         return self._get_orderpoint_action()
 
-    def action_replenish(self):
-        self._procure_orderpoint_confirm(company_id=self.env.company)
+    def action_replenish(self, raise_user_error=True):
+        # Force to recompute because if, for example, a RFQ cancelled after last
+        # compute, the qty_forecast won't be correct.
+        self._compute_qty()
+        self._procure_orderpoint_confirm(company_id=self.env.company, raise_user_error=raise_user_error)
```
Comment (sle-odoo) [inline addons/stock/models/stock_orderpoint.py:216]: @amoyaux will make a different patch (he thinks a depends is missing)

File: addons/stock/models/stock_move.py (hunk 2 above)
Comment (yhu-odoo) [reply addons/stock/models/stock_move.py]: @sle-odoo I moved this to _action_done, action_assign is not called when validate a picking.

File: addons/stock/models/stock_picking.py (hunk 3 above)
Comment (yhu-odoo) [reply addons/stock/models/stock_picking.py]: change applied

File: addons/stock/models/stock_orderpoint.py (hunk 4 above)
Comment (yhu-odoo) [reply addons/stock/models/stock_orderpoint.py]: removed

File: addons/stock/models/stock_picking.py [hunk 5]
```diff
@@ -680,6 +680,17 @@ def action_assign(self):
...
+
+        # run scheduler for moves are not fully reserved
+        for move in moves.filtered(lambda m: m.state in ['confirmed', 'partially_available']):
+            # get all the parents of the move's source location
+            cur_location = move.location_id
+            parent_locations = cur_location
+            while cur_location.location_id:
+                cur_location = cur_location.location_id
+                parent_locations |= cur_location
+            move.product_id.orderpoint_ids.filtered(lambda op: op.location_id in parent_locations and op.trigger == 'auto')._procure_orderpoint_confirm(company_id=move.company_id, raise_user_error=False)
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py:692]: performance issue here

there's two things you should do
1) call _procure_orderpoint_confirm once for all the orderpoint linked to a product
2) get rid of the while .. going up with the parent_id column and use a child_of/parent_of operator

Comment (sle-odoo) [reply addons/stock/models/stock_picking.py:692]: ideally we would a have a data structure containing somehow the grouped products, locations and companies impacted (maybe we could infer the company from the location thoguh)

something similar to this (but the grouping won't work)
``` python
res = defaultdict(lambda: self.env['stock.location'], self.env['res.company'])
for move in moves:
    res[move.product_id] = (move.location_id, move.company_id)
    
for product, (location, company) in res.iteritems():
    orderpoints = self.env['stock.warehouse.orderpoint'].search([
        ('product_id', '=', product.id),
        ('trigger', '=', 'auto'),
        ('location_id', 'parent_of', location.id)
    ])
    orderpoints._procure_orderpoint_confirm(company_id=company, raise_user_error=False)
```
we could have the same orderpoints triggered two times if two stock moves are processed in wh/stock/shelf1 and wh/stock/shelf2, mybe we could in a second times merge these two liens by a sigle wh/stoch

Comment (sle-odoo) [reply addons/stock/models/stock_picking.py:692]: limit=1 on search
store the found orderpoint in a recordset
call _procure_orderpoint_confirm in all orderpoints at once

Comment (amoyaux) [reply addons/stock/models/stock_picking.py:692]:
``` python
orderpoints_by_company = defaultdict(lambda: self.env['stock.location.orderpoint'])
for move in moves:
    orderpoints = self.env['stock.warehouse.orderpoint'].search([
        ('product_id', '=', move.product_id.id),
        ('trigger', '=', 'auto'),
        ('location_id', 'parent_of', move.location_id.id),
        ('company_id', '=', move.company_id.id)
    ], limit=1)
    if orderpoints:
        orderpoints_by_company[move.company_id] |= orderpoints.id
for company, orderpoints in orderpoints_by_company.items():
    orderpoints._procure_orderpoint_confirm(company_id=company, raise_user_error=False)
```

Comment (yhu-odoo) [reply addons/stock/models/stock_picking.py:692]: @amoyaux I think we can remove ('company_id', '=', move.company_id.id) in the search since we already have the location.

Comment (yhu-odoo) [reply addons/stock/models/stock_picking.py]: @sle-odoo @amoyaux Change applied

File: addons/stock/models/stock_picking.py [hunk 6]
```diff
@@ -728,6 +743,15 @@ def _action_done(self):
                     todo_moves |= new_move
         todo_moves._action_done(cancel_backorder=self.env.context.get('cancel_backorder'))
         self.write({'date_done': fields.Datetime.now()})
+
+        # if incoming moves make other confirmed/partially_available moves available, assign them
+        done_incoming_moves = self.filtered(lambda p: p.picking_type_id.code == 'incoming').mapped('move_lines').filtered(lambda m: m.state == 'done')
+        domains = []
+        for move in done_incoming_moves:
+            domains.append([('product_id', '=', move.product_id.id), ('location_id', '=', move.location_dest_id.id)])
+        moves_to_reserve = self.env['stock.move'].search(expression.AND([[('state', 'in', ['confirmed', 'partially_available']), ('procure_method', '=', 'make_to_stock')], expression.OR(domains)]))
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py:752]: sorry to be picky but could you extract the static part of the domain to a variable so that it's 1) clearer 2) the line is shorter ?

File: addons/stock/models/stock_picking.py [hunk 7]
```diff
@@ -728,6 +743,15 @@ def _action_done(self):
                     todo_moves |= new_move
         todo_moves._action_done(cancel_backorder=self.env.context.get('cancel_backorder'))
         self.write({'date_done': fields.Datetime.now()})
+
+        # if incoming moves make other confirmed/partially_available moves available, assign them
+        done_incoming_moves = self.filtered(lambda p: p.picking_type_id.code == 'incoming').mapped('move_lines').filtered(lambda m: m.state == 'done')
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py:748]: the mapped is now useless

File: addons/stock/tests/test_move2.py [hunk 8]
```diff
@@ -2790,3 +2790,134 @@ def test_delay_alert_3(self):
         self.assertFalse(pick.delay_alert_date)
         self.assertFalse(pack.delay_alert_date)
         self.assertFalse(ship.delay_alert_date)
+
+
+class TestAutoAssign(TestStockCommon):
+    def setUp(self):
+        super(TestAutoAssign, self).setUp()
```
Comment (sle-odoo) [inline addons/stock/tests/test_move2.py:2797]: this setup is useless

**Comment by yhu-odoo** [comment]
@sle-odoo all change applied

**Comment by sle-odoo** [comment]
robodoo r+
